gems/aws-sdk-core/lib/aws-sdk-core/plugins/http_checksum.rb

```diff
@@ -0,0 +1,56 @@
+require 'openssl'
+
+module Aws
+  module Plugins
+    # @api private
+    class HttpChecksum < Seahorse::Client::Plugin
+
+      # @api private
+      class Handler < Seahorse::Client::Handler
+        CHUNK_SIZE = 1 * 1024 * 1024 # one MB
+
+        def call(context)
+          if context.operation.http_checksum_required
+            puts "------------i am a checksum required operation"
+            body = context.http_request.body
+            context.http_request.headers['Content-Md5'] ||= md5(body)
+          end
+          @handler.call(context)
+        end
+
+        private
+
+        # @param [File, Tempfile, IO#read, String] value
+        # @return [String<MD5>]
+        def md5(value)
+          if (File === value || Tempfile === value) && !value.path.nil? && File.exist?(value.path)
+            OpenSSL::Digest::MD5.file(value).base64digest
+          elsif value.respond_to?(:read)
+            md5 = OpenSSL::Digest::MD5.new
+            update_in_chunks(md5, value)
+            md5.base64digest
+          else
+            OpenSSL::Digest::MD5.digest(value).base64digest
+          end
+        end
+
+        def update_in_chunks(digest, io)
```

Comment:
I believe there is an opportunity for us to avoid reading the body twice here.  siv4 requires the `x-amz-content-sha256` header to be set and will read the body to compute it.  However, since we are already doing so here, we can compute and set it here as well.

Reply:
Can you link to where that's done? It maybe makes sense to do that here, but if it's part of the sigv4 gem then maybe not.

Reply:
Talked offline - will be doing this later when other modeled checksums are supported.